### PR TITLE
Register executors as components.

### PR DESCRIPTION
**Pull request by vkryachko**
The intent for those is to be used by all Firebase SDKs and forbid creating their own at will.

### CODE REVIEW

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatController.java [hunk 1]
```diff
@@ -118,12 +113,12 @@ private DefaultHeartBeatController(
...
       Set<HeartBeatConsumer> consumers,
-      Provider<UserAgentPublisher> userAgentProvider) {
+      Provider<UserAgentPublisher> userAgentProvider,
+      Executor backgroundExecutor) {
     this(
         () -> new HeartBeatInfoStorage(context, persistenceKey),
         consumers,
-        new ThreadPoolExecutor(
-            0, 1, 30, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), THREAD_FACTORY),
+        backgroundExecutor,
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatController.java:121]: @VinayGuthal is it safe to switch to a multi-threaded executor, looks like this used to rely on a single-threaded one.

File: firebase-common/src/main/java/com/google/firebase/concurrent/CustomThreadFactory.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+package com.google.firebase.concurrent;
+
+import java.util.Locale;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicLong;
+
+class CustomThreadFactory implements ThreadFactory {
+  private static final ThreadFactory DEFAULT = Executors.defaultThreadFactory();
+  private final AtomicLong threadCount = new AtomicLong();
```
Comment (rlazo) [inline firebase-common/src/main/java/com/google/firebase/concurrent/CustomThreadFactory.java:24]: why the counter?

File: firebase-common/src/test/AndroidManifest.xml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2019 Google LLC -->
```
Comment (rlazo) [inline firebase-common/src/test/AndroidManifest.xml:2]:
```suggestion
<!-- Copyright 2022 Google LLC -->
```

File: tools/lint/src/main/kotlin/CheckRegistry.kt [hunk 4]
```diff
@@ -29,6 +29,8 @@ class CheckRegistry : IssueRegistry() {
                 NonAndroidxNullabilityDetector.NON_ANDROIDX_NULLABILITY,
                 DeferredApiDetector.INVALID_DEFERRED_API_USE,
                 ProviderAssignmentDetector.INVALID_PROVIDER_ASSIGNMENT
+                // TODO(vkryachko): enable the check after migration.
```
Comment (rlazo) [inline tools/lint/src/main/kotlin/CheckRegistry.kt:32]: You mean, after adding the @suppressLint in current usages?

File: firebase-common/src/main/java/com/google/firebase/concurrent/CustomThreadFactory.java (hunk 2 above)
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/concurrent/CustomThreadFactory.java:24]: This is common practice to give different names to threads, i.e. `executor-foo-1`, `executor-foo-2` etc.

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatController.java (hunk 1 above)
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatController.java:121]: Discussed offline, it should be safe

File: firebase-common/src/main/java/com/google/firebase/concurrent/CustomThreadFactory.java (hunk 2 above)
Comment (rlazo) [reply firebase-common/src/main/java/com/google/firebase/concurrent/CustomThreadFactory.java:24]: Didn't see it in the name, thanks!

**Review by rlazo — APPROVED** [review]

**Pull request merged**